off_chain_data/application-go/parser/endorserTransaction.go

```diff
@@ -0,0 +1,115 @@
+package parser
+
+import (
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset"
+	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
+	"google.golang.org/protobuf/proto"
+)
+
+type endorserTransaction struct {
+	transaction   *peer.Transaction
+	readWriteSets func() ([]*readWriteSet, error)
+}
+
+func parseEndorserTransaction(transaction *peer.Transaction) *endorserTransaction {
+	result := &endorserTransaction{transaction, nil}
+	result.readWriteSets = sync.OnceValues(result.unmarshalReadWriteSets)
+	return result
+}
+
+func (p *endorserTransaction) unmarshalReadWriteSets() ([]*readWriteSet, error) {
+	chaincodeActionPayloads, err := p.unmarshalChaincodeActionPayloads()
+	if err != nil {
+		return nil, err
+	}
+
+	chaincodeEndorsedActions, err := p.extractChaincodeEndorsedActionsFrom(chaincodeActionPayloads)
+	if err != nil {
+		return nil, err
+	}
+
+	proposalResponsePayloads, err := p.unmarshalProposalResponsePayloadsFrom(chaincodeEndorsedActions)
+	if err != nil {
+		return nil, err
+	}
+
+	chaincodeActions, err := p.unmarshalChaincodeActionsFrom(proposalResponsePayloads)
+	if err != nil {
+		return nil, err
+	}
+
+	txReadWriteSets, err := p.unmarshalTxReadWriteSetsFrom(chaincodeActions)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseReadWriteSets(txReadWriteSets), nil
+}
+
+func (p *endorserTransaction) unmarshalChaincodeActionPayloads() ([]*peer.ChaincodeActionPayload, error) {
+	result := []*peer.ChaincodeActionPayload{}
+	for _, transactionAction := range p.transaction.GetActions() {
+		chaincodeActionPayload := &peer.ChaincodeActionPayload{}
+		if err := proto.Unmarshal(transactionAction.GetPayload(), chaincodeActionPayload); err != nil {
+			return nil, err
+		}
+
+		result = append(result, chaincodeActionPayload)
+	}
+	return result, nil
+}
+
+func (*endorserTransaction) extractChaincodeEndorsedActionsFrom(chaincodeActionPayloads []*peer.ChaincodeActionPayload) ([]*peer.ChaincodeEndorsedAction, error) {
+	result := []*peer.ChaincodeEndorsedAction{}
```

Comment:
This is really a nitpick... I think the result slice you create here (or rather its underlying array) is going to get discarded as soon as you append to it in the loop since there isn't capacity to append more elements. You can save the work of creating storage that will be discarded by using a nil value, which still works as an empty slice:

```go
var result []*peer.ChaincodeEndorsedAction
```

Or you can make a slice with capacity to hold `len(chaincodeActionPayloads)` elements up-front.

Same comment for the other instances of creating an empty (zero capacity) slice that you will then append to in this file.

Like I say, it's a really minor detail. Change it only if you want to.